src/main/java/net/minecraftforge/client/gui/config/ControlCreator.java

```diff
@@ -0,0 +1,525 @@
+            else if (value instanceof Double)
+                createTextualNumericInteractionWidget((Interactor<Double>) interactor, Double::parseDouble, getLongestValueStringLength(-Double.MAX_VALUE));
+            else
+                throw new IllegalStateException("Don't know how to make a widget for Non-ranged Number " + value);
+        }
+    }
+
+    /**
+     * @return null if a slider should not be created for the interactor.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T extends Number, C extends Number & Comparable<? super C>> Range<C> getRangeForSliderCreation(Interactor<T> interactor) {
+        Range<C> range = (Range<C>) interactor.getData(RANGE_KEY);
+        if (range == null)
+            return null;
+        double min = range.getMin().doubleValue();
+        double max = range.getMax().doubleValue();
+        if (max - min > 10)
+            return null;
+        return range;
+    }
+
+    public int getLongestValueStringLength(Number value) {
+        return Objects.toString(value).length();
+    }
+
+    public <T extends Number> void createSliderNumericInteractionWidget(Interactor<T> interactor, Range<? extends Number> range, Function<Double, T> converter) {
+        Slider.ISlider responder = slider -> {
+            Double sliderValue = slider.getValue();
+            T newValue = converter.apply(sliderValue);
+            boolean isValid = interactor.isValid(newValue);
+            if (isValid)
+                interactor.save(newValue);
+            interactor.onUpdate(isValid, newValue);
+        };
+
+        interactor.control = createSlider(interactor, range, responder);
+
+        interactor.addUpdateResponder((isValid, newValue) -> {
+            Slider control = ((Slider) interactor.control);
+            if (!Objects.equals(control.getValue(), newValue.doubleValue())) {
+                control.setValue(newValue.doubleValue());
+                control.updateSlider();
+            }
+        });
+    }
+
+    public <T extends Number> void createTextualNumericInteractionWidget(Interactor<T> interactor, Function<String, T> parser, int longestValueStringLength) {
+        TextFieldWidget textField = createTextField(interactor);
+        interactor.control = textField;
+
+        textField.setMaxStringLength(longestValueStringLength);
+        MutableObject<String> currentlyRespondingTo = new MutableObject<>();
+        textField.setResponder(newText -> {
+            if (Objects.equals(currentlyRespondingTo.getValue(), newText))
+                // We are being called from inside visualsUpdater, don't recurse & stackoverflow
+                return;
+            T newValue;
+            try {
+                newValue = parser.apply(newText);
+            } catch (NumberFormatException ignored) {
+                ((TextFieldWidget) interactor.control).setTextColor(RED);
```

Comment:
Shouldn't this get a red border? In case the text is empty

Reply:
Yeah, that would be much better solution than changing the text color. I’ll see if I can do that. If I can I’ll also do it for buttons - the text color is super ugly IMO